### PR TITLE
[TOSA] Usage of 32bit integer for 'index to float' in rfft2d

### PR DESCRIPTION
**Pull request by d-smirnov**
Lowering of rfft2d to linalg now uses index to i32 cast if an output float is of 32bit and cast to i64 otherwise. 
@eric-k256 @GeorgeARM

### CODE REVIEW

**Review by eric-k256 — APPROVED** [review]
looks okay to me.

**Comment by GeorgeARM** [comment]
Thanks for this PR. Overall it looks fine to me.
Could you please elaborate in your commit-msg what is the reasoning behind doing this?

If this is due to conversion targets not being able to handle 64bit integer arithmetic (e.g. SPIRV) then I think that this might require more discussion. Not sure how we can handle or if we need a more immediate casting operator from index to floating point that can be lowered later on as per target needs.

**Comment by MaheshRavishankar** [comment]
Thanks for the tag @GeorgeARM . I am on vacation right now but @antiagainst or @kuhar can help here

**Review by RoboTux — COMMENTED** [review]
LGTM.

File: mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp [hunk 1]
```diff
@@ -2284,8 +2284,11 @@ struct RFFT2dConverter final : public OpRewritePattern<RFFT2dOp> {
 
   static Value castIndexToFloat(OpBuilder &builder, Location loc,
                                 FloatType type, Value value) {
-    auto integerVal =
-        builder.create<arith::IndexCastUIOp>(loc, builder.getI64Type(), value);
+    auto integerVal = builder.create<arith::IndexCastUIOp>(
+        loc,
+        32 < type.getIntOrFloatBitWidth() ? builder.getI64Type()
```
Comment (kuhar) [inline mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp:2289]: nit: it's more common to see constants as the RHS of comparison operators

Comment (kuhar) [reply mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp:2289]: Also, should we use an integer type of the same bitwidth? For example `i16` when casting to `f16`.

**Comment by kuhar** [comment]
> If this is due to conversion targets not being able to handle 64bit integer arithmetic "easily" or at all (e.g. SPIRV) then I think that this might require more discussion. Not sure how we can handle or if we need a more immediate casting operator from index to floating point that can be lowered later on as per target needs.

FYI, I have patterns designed with this set of constraints in mind: https://github.com/llvm/llvm-project/blob/32532c2bbedc6e15eb3eae65cb6049e196cef9ad/mlir/lib/Dialect/Arith/Transforms/IntNarrowing.cpp#L396-L423

**Comment by d-smirnov** [comment]
Yep, the goal of the patch is to omit `index -> i64` conversion if possible
in lowering rfft2d Op as the subsequent transformation to SPIRV not able to
"easily'' handle 64bit integer arithmetic (which seems excessive in this
particular case anyway).

--
Dmitriy Smirnov

File: mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp (hunk 1 above)
Comment (d-smirnov) [reply mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp:2289]: tosa.rfft2d defined only for f32, probably we should not bother unless there is some strong case for this.

Comment (GeorgeARM) [reply mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp:2289]: I agree with @kuhar . constant on RHS are more aligned. Could you please change?

**Comment by d-smirnov** [comment]
PR was accidentally closed. Reopened now.

**Comment by GeorgeARM** [comment]
LGTM

**Review by GeorgeARM — APPROVED** [review]

**Pull request merged**